node/network/bridge/src/validator_discovery.rs

```diff
@@ -115,39 +114,93 @@ impl NonRevokedConnectionRequestState {
 	}
 }
 
+/// Will be called by [`Service::on_request`] when a request was revoked.
+///
+/// Takes the `map` of requested validators and the `id` of the validator that should be revoked.
+///
+/// Returns `Some(id)` iff the request counter is `0`.
+fn on_revoke(map: &mut HashMap<AuthorityDiscoveryId, u64>, id: AuthorityDiscoveryId) -> Option<AuthorityDiscoveryId> {
+	if let hash_map::Entry::Occupied(mut entry) = map.entry(id) {
+		if entry.get_mut().saturating_sub(1) == 0 {
+			return Some(entry.remove_entry().0);
+		}
+	}
+
+	None
+}
+
+fn peer_id_from_multiaddr(addr: &Multiaddr) -> Option<PeerId> {
+	addr.iter().last().and_then(|protocol| if let Protocol::P2p(multihash) = protocol {
+		PeerId::from_multihash(multihash).ok()
+	} else {
+		None
+	})
+}
 
 pub(super) struct Service<N, AD> {
-	// we assume one PeerId per AuthorityId is enough
-	connected_validators: HashMap<AuthorityDiscoveryId, PeerId>,
-	// the `u64` counts the number of pending non-revoked requests for this validator
+	// Peers that are connected to us and authority ids associated to them.
+	connected_peers: HashMap<PeerId, HashSet<AuthorityDiscoveryId>>,
+	// The `u64` counts the number of pending non-revoked requests for this validator
 	// note: the validators in this map are not necessarily present
 	// in the `connected_validators` map.
 	// Invariant: the value > 0 for non-revoked requests.
 	requested_validators: HashMap<AuthorityDiscoveryId, u64>,
 	non_revoked_discovery_requests: Vec<NonRevokedConnectionRequestState>,
 	// PhantomData used to make the struct generic instead of having generic methods
-	network: PhantomData<N>,
-	authority_discovery: PhantomData<AD>,
+	_phantom: PhantomData<(N, AD)>,
 }
 
 impl<N: Network, AD: AuthorityDiscovery> Service<N, AD> {
 	pub fn new() -> Self {
 		Self {
-			connected_validators: HashMap::new(),
+			connected_peers: HashMap::new(),
 			requested_validators: HashMap::new(),
 			non_revoked_discovery_requests: Vec::new(),
-			network: PhantomData,
-			authority_discovery: PhantomData,
+			_phantom: PhantomData,
+		}
+	}
+
+	/// Find connected validators using the given `validator_ids`.
+	///
+	/// Returns a [`HashMap`] that contains the found [`AuthorityDiscoveryId`]'s and their associated [`PeerId`]'s.
+	async fn find_connected_validators(
+		&mut self,
+		validator_ids: &[AuthorityDiscoveryId],
+		authority_discovery_service: &mut AD,
+	) -> HashMap<AuthorityDiscoveryId, PeerId> {
+		let mut result = HashMap::new();
+
+		for id in validator_ids {
+			// First check if we already cached the validator
+			if let Some(pid) = self.connected_peers
+				.iter()
```

Comment:
this is going to be `validator_ids.len()` X `self.connected_peers.len()`, probably not an issue if `validator_ids.len()` is small

Reply:
I know and by default we are connected to like 50 nodes? I think this is fine. But we can also split this up into a hashmap and hashset.